lib/spack/spack/solver/asp.py

```diff
@@ -774,6 +774,34 @@ def package_compiler_defaults(self, pkg):
                 pkg.name, cspec.name, cspec.version, -i * 100
             ))
 
+    def cfg_required_rules(self, pkg):
+        experimental_hack_selector = 1
+        if experimental_hack_selector == 1:
+            # The goal here is to say "if we need to build MPI, then it has to
+            # either satisfy @4.0.2+static or @4.0.1~static"
+            test_rules = [
+                ("openmpi", ["@4.0.2+static", "@4.0.1~static"])
+            ]
+        elif experimental_hack_selector == 2:
+            # Example of combinatoric constraints, you can mix either compiler
+            # with either version, but it must be one of the 4 combinations
+            test_rules = [
+                ("openmpi", ["%gcc", "%clang"]),
+                ("openmpi", ["@4.0.1", "@4.0.0"])
+            ]
+        stanza_id = 0
+        member_id = 100000
+        for pkg_name, stanza in test_rules:
+            self.gen.fact(fn.condition_group(pkg_name, stanza_id))
+            for spec_str in stanza:
+                spec, = spack.cmd.parse_specs(spec_str)
+                if not spec.name:
+                    spec.name = pkg_name
+                self.gen.fact(fn.condition_group_member(member_id, stanza_id))
+                self.impose(member_id, spec, name=pkg_name, node=False)
+                member_id += 1
+            stanza_id += 1
```

Comment:
Nice! It looks like this is on the right track

I have not had a chance to think about the structure of this and the ASP in detail, but it'll need a rework for your TODO below anyway.  So I suggest working with @becker33 on that part.

I don't think this needs to be integrated with spec_clauses, but I could be wrong -- I think the way you've done it here, with `impose()` is good.  I don't fully understand the need for `stanza_id` *and* `member_id` -- but you will need a condition id for the `impose()`.  I think the condition can just have a single requirement -- `node(pkg_name)` and you can simplify what you have here.  @becker33 and @alalazo should be able to help with that.

Reply:
The member_id is the condition ID for `impose`, the `stanza_id` is the identifier for the group of conditions. I think it is minimal (without an identifier for the group, how can I say "one of the constraints in this group must apply"?).